### PR TITLE
Add indexer

### PR DESCRIPTION
**Pull request by classicalliu**

### CODE REVIEW

File: crates/indexer/src/helper.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+impl PolyjuiceArgs {
+    // https://github.com/nervosnetwork/godwoken-polyjuice/blob/v0.6.0-rc1/polyjuice-tests/src/helper.rs#L322
+    pub fn decode(args: &[u8]) -> anyhow::Result<Self> {
+        let is_create = args[7] == 3u8;
+        let gas_limit = u64::from_le_bytes(args[8..16].try_into()?);
+        let gas_price = u128::from_le_bytes(args[16..32].try_into()?);
+        let value = u128::from_le_bytes(args[32..48].try_into()?);
+        let input_size = u32::from_le_bytes(args[48..52].try_into()?);
+        let input: Vec<u8> = args[52..(52 + input_size as usize)].to_vec();
```
Comment (quake) [inline crates/indexer/src/helper.rs:28]: Potential out of bound attacks, please check the args len and input_size first

File: crates/indexer/src/helper.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        GW_LOG_POLYJUICE_USER => {
+            let mut offset: usize = 0;
+            let mut address = [0u8; 20];
+            address.copy_from_slice(&data[offset..offset + 20]);
+            offset += 20;
+            let mut data_size_bytes = [0u8; 4];
+            data_size_bytes.copy_from_slice(&data[offset..offset + 4]);
+            offset += 4;
+            let data_size: u32 = u32::from_le_bytes(data_size_bytes);
+            let mut log_data = vec![0u8; data_size as usize];
```
Comment (quake) [inline crates/indexer/src/helper.rs:155]: Potential memory attacks, please check data_size before allocating a vector

File: crates/indexer/src/helper.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+            let mut topics_count_bytes = [0u8; 4];
+            topics_count_bytes.copy_from_slice(&data[offset..offset + 4]);
+            offset += 4;
+            let topics_count: u32 = u32::from_le_bytes(topics_count_bytes);
+            let mut topics = Vec::new();
+            log::debug!("topics_count: {}", topics_count);
+            for _ in 0..topics_count {
+                let mut topic = [0u8; 32];
+                topic.copy_from_slice(&data[offset..offset + 32]);
```
Comment (quake) [inline crates/indexer/src/helper.rs:168]: Potential out of bound attacks, please check the offset against data length first

File: crates/indexer/src/runner.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        let row: Option<(Decimal,)> =
+            sqlx::query_as("select number from blocks order by number desc limit 1;")
+                .fetch_optional(&self.pg_pool)
+                .await?;
+        let num = row.and_then(|(n,)| n.to_u64());
+        Ok(num)
+    }
+
+    pub async fn insert(&mut self) -> Result<bool> {
+        let local_tip = self.tip().await?;
```
Comment (quake) [inline crates/indexer/src/runner.rs:72]: This fn doesn't seem to handle the layer 2 block forking case

**Review by quake — CHANGES_REQUESTED** [review]

File: crates/indexer/src/helper.rs (hunk 2 above)
Comment (jjyr) [reply crates/indexer/src/helper.rs:155]: @Flouse Do we have max size of this data?

File: crates/indexer/src/helper.rs (hunk 1 above)
Comment (Flouse) [reply crates/indexer/src/helper.rs:28]: `GW_MAX_L2TX_ARGS_SIZE`, known as the max size of `PolyjuiceArgs` is 128KB.

It is defined in [`godwoken-scripts/c/gw_def.h`](https://github.com/nervosnetwork/godwoken-scripts/blob/fb69b35aaadef70dba162331bec3e2f25517714f/c/gw_def.h).

File: crates/indexer/src/helper.rs (hunk 2 above)
Comment (Flouse) [reply crates/indexer/src/helper.rs:155]: According to [Ethereum Yellow Paper (Page 5)](https://ethereum.github.io/yellowpaper/paper.pdf), there is no size limit to the log_data.

But, as you know, Polyjuice is running on CKB_VM.
Only about 1MB heap memory can be [malloced](https://github.com/nervosnetwork/godwoken-polyjuice/blob/main/c/polyjuice.h#L636) in CKB_VM.

Comment (jjyr) [reply crates/indexer/src/helper.rs:155]: I think a reasonable value is 4K or 8K, we can revisit this value if data_size is longer than this.

File: crates/indexer/src/error_receipt_indexer.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            .bind(hex(record.tx_hash.as_slice())?)
+            .bind(Decimal::from(record.block_number))
+            .bind(Decimal::from(record.cumulative_gas_used))
+            .bind(Decimal::from(record.gas_used))
+            .bind(Decimal::from(record.status_code))
+            .bind(record.status_reason)
+            .execute(&mut db)
+            .await?;
+
+        db.commit().await?;
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/error_receipt_indexer.rs]: updated

File: crates/indexer/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+version = "0.1.0"
+authors = ["Nervos Network"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+gw-types = { git = "https://github.com/nervosnetwork/godwoken.git", tag = "v0.6.9-rc1" }
+gw-common = { git = "https://github.com/nervosnetwork/godwoken.git", tag = "v0.6.9-rc1" }
+gw-mem-pool = { git = "https://github.com/nervosnetwork/godwoken.git", tag = "v0.6.9-rc1" }
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/Cargo.toml]: updated

File: crates/indexer/src/error_receipt_indexer.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        match gw_log {
+            GwLog::PolyjuiceSystem {
+                gas_used,
+                cumulative_gas_used,
+                created_address: _,
+                status_code,
+            } => {
+                let isnt_string = |t: &ethabi::token::Token| -> bool {
+                    !matches!(t, ethabi::token::Token::String(_))
+                };
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/error_receipt_indexer.rs]: updated

File: crates/indexer/src/helper.rs (hunk 1 above)
Comment (classicalliu) [reply crates/indexer/src/helper.rs]: updated

File: crates/indexer/src/helper.rs (hunk 2 above)
Comment (classicalliu) [reply crates/indexer/src/helper.rs:184]: Will throw an error if data size greater than 4KB.

File: crates/indexer/src/helper.rs (hunk 3 above)
Comment (classicalliu) [reply crates/indexer/src/helper.rs:200]: updated

File: crates/indexer/src/runner.rs (hunk 4 above)
Comment (classicalliu) [reply crates/indexer/src/runner.rs:72]: Yes, will draft a new pr to do this.

File: crates/indexer/src/helper.rs (hunk 3 above)
Comment (Flouse) [reply crates/indexer/src/helper.rs:200]: Solidity does allow up to 4 topics (the maximum supported by the EVM).
> Events are an abstraction of the Ethereum logging/event-watching protocol. Log entries provide the contract’s address, a series of up to four topics and some arbitrary length binary data. 
> https://docs.soliditylang.org/en/latest/abi-spec.html?highlight=log#events

File: crates/indexer/src/error_receipt_indexer.rs [hunk 8]
```diff
@@ -116,14 +116,14 @@ impl From<ErrorTxReceipt> for ErrorReceiptRecord {
...
-                    !matches!(t, ethabi::token::Token::String(_))
+                let is_string = |t: &ethabi::token::Token| -> bool {
+                    matches!(t, ethabi::token::Token::String(_))
                 };
 
                 // First 4 bytes are func signature
                 let status_reason =
                     match ethabi::decode(&[ethabi::ParamType::String], &receipt.return_data[4..]) {
-                        Ok(tokens) if !tokens.iter().any(isnt_string) => {
+                        Ok(tokens) if !tokens.iter().all(is_string) => {
```
Comment (quake) [inline crates/indexer/src/error_receipt_indexer.rs]: typo?
```
Ok(tokens) if tokens.iter().all(is_string)
```

File: crates/indexer/src/pool.rs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+        smol::block_on(async {
+            let mut opts: PgConnectOptions = indexer_config.pg_url.parse()?;
+            opts.log_statements(log::LevelFilter::Debug)
+                .log_slow_statements(log::LevelFilter::Warn, Duration::from_secs(5));
+            PgPoolOptions::new()
+                .max_connections(5)
+                .connect_with(opts)
+                .await
+        }).unwrap()
```
Comment (quake) [inline crates/indexer/src/pool.rs:22]: suggest to use `now_or_never` without runtime poll

```suggestion
        let mut opts: PgConnectOptions = indexer_config.pg_url.parse().expect("...");
        opts.log_statements(log::LevelFilter::Debug).log_slow_statements(log::LevelFilter::Warn, Duration::from_secs(5));

        PgPoolOptions::new().max_connections(5).connect_with(opts).now_or_never().expect("...")

File: crates/indexer/src/error_receipt_indexer.rs [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            })
+            .detach();
+        }
+
+        smol::spawn(async move {
+            if let Err(err) = Self::insert_error_tx_receipt(receipt).await {
+                log::error!("insert error tx receipt {}", err);
+            }
+        })
+        .await;
```
Comment (quake) [inline crates/indexer/src/error_receipt_indexer.rs:79]: no need to spawn a future and await

```suggestion
        if let Err(err) = Self::insert_error_tx_receipt(receipt).await {
            log::error!("insert error tx receipt {}", err);
        }
```

**Review by quake — APPROVED** [review]

File: crates/indexer/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+fn main() -> Result<()> {
+    env_logger::init_from_env(env_logger::Env::default().default_filter_or("info"));
+    let indexer_config = read_indexer_config("./indexer-config.toml")?;
+
+    let listener_task = start_listen_error_tx_receipt(&indexer_config);
+
+    let mut runner = Runner::new(indexer_config)?;
+
+    smol::block_on(runner.run())?;
```
Comment (jjyr) [inline crates/indexer/src/main.rs:15]: This line prevents the execution of the next line.

We should spawn tasks first, then wait for them.

```
let runner_task = smol::spawn(..);
let listener_task = smol::spawn(..);
smol::block_on(async{
  runner_task.await?;
  listener_task.await?;
});
```

**Review by jjyr — CHANGES_REQUESTED** [review]

**Review by Flouse — APPROVED** [review]

**Review by jjyr — APPROVED** [review]
